### PR TITLE
[libc][math][c23] Add sinf16 C23 math function

### PR DESCRIPTION
**Pull request by wldfngrs**
- Implementation of sin for 16-bit floating point inputs
- Exhaustive tests for `sinf16()`

### CODE REVIEW

**Comment by wldfngrs** [comment]
@lntue please review

File: libc/src/math/generic/sinf16.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
+//===-- Half-precision sin function ------------------------------------===//
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:1]: Nit: formatting.

```suggestion
//===-- Half-precision sin(x) function ------------------------------------===//
```

File: libc/test/src/math/sinf16_test.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+//===-- Exhaustive test for sinf16 ----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===---------------------------------------------------------------------===//
```
Comment (overmighty) [inline libc/test/src/math/sinf16_test.cpp:7]: Nit: formatting.

```suggestion
//===----------------------------------------------------------------------===//
```

File: libc/src/math/generic/sinf16.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+constexpr fputil::ExceptValues<float16, N_EXCEPTS> SINF16_EXCEPTS{{
+    // (input, RZ output, RU offset, RD offset, RN offset)
+    {0x585c, 0x3ba3, 1, 0, 1},
+    {0x5cb0, 0xbbff, 0, 1, 0},
+    {0x51f5, 0xb80f, 0, 1, 0},
+}};
+
+LLVM_LIBC_FUNCTION(float16, sinf16, (float16 x)) {
+  using FPBits = typename fputil::FPBits<float16>;
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:32]: `typename` is redundant here.

```suggestion
  using FPBits = fputil::FPBits<float16>;
```

File: libc/src/math/generic/sinf16.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  FPBits xbits(x);
+
+  uint16_t x_u = xbits.uintval();
+  uint16_t x_abs = x_u & 0x7fff;
+  float xf = x;
+
+  // Range reduction:
+  // For !x| > pi/32, we perform range reduction as follows:
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:40]:
```suggestion
  // For |x| > pi/32, we perform range reduction as follows:
```

File: libc/src/math/generic/sinf16.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  //   y = x * 32/pi - k
+  //
+  // Once k and y are computed, we then deduce the answer by the sine of sum
+  // formula:
+  //   sin(x) = sin((k + y) * pi/32)
+  //   	      = sin(k * pi/32) * cos(y * pi/32) +
+  //   	        sin(y * pi/32) * cos(k * pi/32)
+
+  // Handle exceptional values
+  if (LIBC_UNLIKELY(x_abs == 0x585C || x_abs == 0x5cb0 || x_abs == 0x51f5)) {
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:56]: Nit: stick to lowercase hexadecimal for consistency.

```suggestion
  if (LIBC_UNLIKELY(x_abs == 0x585c || x_abs == 0x5cb0 || x_abs == 0x51f5)) {
```

File: libc/src/math/generic/sinf16.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    bool x_sign = x_u >> 15;
+    if (auto r = SINF16_EXCEPTS.lookup_odd(x_abs, x_sign);
+        LIBC_UNLIKELY(r.has_value()))
+      return r.value();
+  }
+
+  int rounding = fputil::quick_get_round();
+  if (LIBC_UNLIKELY(x_abs <= 0x13d0)) {
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:64]: Could you add comments for all these cases from here to line 83?

File: libc/src/math/generic/sinf16.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  if (LIBC_UNLIKELY(x_abs == 0x2b45)) {
+    if (rounding == FE_DOWNWARD && xbits.is_neg()) {
+      x_u--;
+      return FPBits(x_u).get_val();
+    }
+  }
+
+  if (LIBC_UNLIKELY(x_abs >= 0x7c00)) {
+    // If value is equal to infinity
+    if (x_abs == 0x7c00) {
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:87]: Could use `xbits.is_inf_or_nan()` and `xbits.is_inf()` for readability.

**Review by overmighty — CHANGES_REQUESTED** [review]

File: libc/src/math/generic/sinf16.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    if (auto r = SINF16_EXCEPTS.lookup_odd(x_abs, x_sign);
+        LIBC_UNLIKELY(r.has_value()))
+      return r.value();
+  }
+
+  // Exhaustive tests show that for |x| <= 0x13d0, 1ULP rounding errors occur.
+  // To fix this, the following apply:
+  // - When x >= 0, and rounding upward, sin(x) == x.
+  // - When x < 0, and rounding downward, sin(x) == x.
+  // - When x < 0, and rounding upward, sin(x) == (x - 1ULP)
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:69]: Nit: I think writing what `0x13d0` corresponds to in hex float notation would be more helpful, and each "When ..., sin(x) = ..." comment should be put right above the corresponding `if` statement.

File: libc/src/math/generic/sinf16.cpp [hunk 9]
```diff
@@ -62,20 +62,21 @@ LLVM_LIBC_FUNCTION(float16, sinf16, (float16 x)) {
...
+  // occur. To fix this, the following apply:
   if (LIBC_UNLIKELY(x_abs <= 0x13d0)) {
+    // When x >= 0, and rounding upward, sin(x) == x.
     if (LIBC_UNLIKELY(x_abs == 0U))
       return x;
 
+    // When x < 0, and rounding downward, sin(x) == x.
     if ((rounding == FE_UPWARD && xbits.is_pos()) ||
         (rounding == FE_DOWNWARD && xbits.is_neg()))
       return x;
```
Comment (overmighty) [inline libc/src/math/generic/sinf16.cpp:77]: The comments don't seem to match the code here.

```suggestion
    // sin(+/-0) = +/-0
    if (LIBC_UNLIKELY(x_abs == 0U))
      return x;

    // When x > 0, and rounding upward, sin(x) == x.
    // When x < 0, and rounding downward, sin(x) == x.
    if ((rounding == FE_UPWARD && xbits.is_pos()) ||
        (rounding == FE_DOWNWARD && xbits.is_neg()))
      return x;
```

**Review by overmighty — APPROVED** [review]

**Review by overmighty — CHANGES_REQUESTED** [review]

**Review by overmighty — APPROVED** [review]

**Pull request merged**